### ISSUE TITLE
Remove public IPs from machines where it is not explicitly needed

### ISSUE THREAD

**Issue by liquidat** [ansible/workshops#448]
##### SUMMARY

In the RHEL workshop, all nodes have public IPs. However, this is not needed: all interaction takes place via the control hosts, so no public IPs are required.

These machine are affected:

- node1
- node2
- node3

##### ISSUE TYPE

 - Bug Report

**Comment by Valien** [comment]
In the labs that I have run we've used the Public IP in order to help validate the http server installation of Apache. I know the lab says to use curl but many times students like the visual look of refreshing a browser and seeing the results as soon as they run the playbook.

**Comment by liquidat** [comment]
@Valien That is a good concern. We have to find the right balance between usability and security.
Any idea if there would be another way to provide this usability to customers without public IPs? Command line browsers are not an option I guess?

**Comment by Valien** [comment]
@liquidat  hmm, yeah, CLI browsers would not be a good option. I've noticed in some of the workshops I've done that adding more complexity to the CLI can really confuse some participants.

My thought because these are workshop systems and their TTL is very short (I use the RHPDS to provision and they are only up for 48 hours and then RHPDS auto-deletes them) I'm not as overly concerned with the security footprint issue. Any SSH work is done using the internal IP from the control node and AFAIK the only external ports/access is 80 for the browser. Granted if someone provisions these by themselves outside of RHPDS they might have ingress set for 0.0.0.0/0 for all but maybe we can document that only ports 80 are open externally and 22 internally?

**Comment by dmc5179** [comment]
At the very least the public IPs shouldn't be used as part of the inventory. The tower node can access the hosts using private IP space. If the machines are powered on/off all of the EIPs change and the entire lab is broken. If the inventory uses the private IP space which is far less likely to change, then a power cycle requires only fixing 1 EIP/DNS entry for each student. So in a workshop with 25 people, like what I did, you would only have to fix 25 hosts and not 100.

Also don't turn off the workshop hosts or this whole workshop collapses

**Comment by cigamit** [comment]
Public IPs and external SSH access are technically required during provisioning, otherwise Ansible can't connect to the VM to set it up.

**Comment by dmc5179** [comment]
@cigamit I called out the inventory. The inventory used by tower to communicate with the nodes should not use the EIP. There's no technical reason for it. The tower node can communicate over the private address space. The EIPs of the nodes doesn't really matter once the workshop is provisioned. If someone really needs access they can jump from the tower node SSH to the other nodes. The fewer EIPs are used as part of the workshop, not the provisioning of the workshop, the less brittle the workshop would be. I'm not sure it's a great idea to provision 4 nodes per user and be unable to turn them off/on without fixing them all.

**Comment by cigamit** [comment]
Yes, I understand that fully, and in the Windows Workshop we use the internal IP for all internal DNSing.

But unless we are going to write the provisioner in a way that will then remove the Public IP / Firewall rules after the workshop is completely setup, then its a moot point, as you need it during setup.  Removing the Public IPs afterwards would make fixing or expanding the workshop difficult (a lot of us don't use RHPDS and sometimes have to add an extra 10 students after initial provisioning).

**Comment by dmc5179** [comment]
What is a moot point? I'm not entirely sure people read what I write. I never said remove the EIPs from any of the nodes. All I said was that the INVENTORY file used on each ansible/tower node to communicate with the other nodes should use the private IP space.

For every student who is added, 4 machines are created, each with an EIP. If I turn off the machines and turn them back on, all 4 EIPs change. The workshop as it stands today uses the EIPs in the inventory which means turning the machines off breaks the entire workshop. If the inventory used the private IP space instead of EIPs, turning the machines off and back on wouldn't cause so much damage. I would then only need to realign the Route53 DNS name of each ansible tower node with its new EIP to fix the workshop.

With the workshop the way it is we have to leave every machine on from provision through to workshop end. That gets pricey with 25 students. I'm not sure how others do it but I don't provision the lab the day before. I try to set it up at least a few days early. I suppose this explains why T2 instances were chosen since all of the days the machines on will rack up CPU credits for use during the lab.

Just to reiterate, I never said get rid of the EIPs. I suggested what should be the easiest change imaginable, use private IPs in the inventory instead of EIPs.

At this point I think I'll just let it go and accept a workshop was made that can't be turned off and back on...

**Comment by cigamit** [comment]
My first statement was regarding the original Issue by the OP.  My second was regarding your statement of "The EIPs of the nodes doesn't really matter once the workshop is provisioned".

As you also stated, the EIPs do matter because of the Route 53 addresses.  They also matter because of the instructor inventory that can currently be used to expand / fix the workshop on the fly by rerunning the provisioner or by adhoc commands.

A side effect of this is that, yes, you can not shutdown the VMs, but thats a minor loss for VMs that are such short lived.

**Comment by dmc5179** [comment]
@cigamit In all these comments you've yet to respond to using private IPs in the INVENTORY file on each students tower node. Not once. Seems odd.

Also, put a giant warning on the workshop readme, don't turn me off or FAIL. I'm sure it's obvious to someone who has done the workshop many times but maybe not for anyone provisioning it for the first time.

**Comment by IPvSean** [comment]
![calm down everyone](https://media.tenor.com/images/c70d1f4273a86101c52369dd12de7bd1/tenor.gif)

A lot of the workshops based their inventory off of a old network workshop where the VMs live in different VPCs, so the history is that they needed the public IPs to work correctly.  In the case of some of the workshops that all live in one VPC they could be turned off for the local inventory on the control node, but its not intentional probably, they just copy paste and use shitty inventories from an old version that just worked.

Jimmy is correct that most if not all workshops use the public IPs for configuration management so we can't really turn them off unless we configure a proxy through the control node for every lab.

> Also, put a giant warning on the workshop readme, don't turn me off or FAIL. I'm sure it's obvious to someone who has done the workshop many times but maybe not for anyone provisioning it for the first time.
This is something that is inherit of AWS and I am not sure how much we should document normal AWS behavior.  The workshops are meant to be turned on for a period of time and then turned off.  If you draft up the warning that would help you, and link to relevant AWS docs I accept PRS.

**Comment by cigamit** [comment]
Actually, I mentioned that we use the internal IPs on the Windows Workshop.  It doesn't matter to me what you use on the others, so there was no reason to give my opinion on the matter.  Also thats not what this Issue was created for, so it's a side point.

Anyway, I would consider this issue closed as we are not going to be able to rid the workshops of the EIPs.

**Comment by IPvSean** [comment]
oh the actual problem is the turning off / turning on of instances in `aws`.  I have not tested recently... so the private_ips remain the same on reboots? @dmc5179 

I have rewrote some parts of the provisioner so honestly just using private IPs between control host to web nodes is not that big of a deal...

I got to do some testing, I like having the public IPs so that folks can actually use their web browser as @Valien has suggested...

**Comment by dmc5179** [comment]
@IPvSean Yes, the private IP address will remain the same through reboots. Any comms between the control hosts and nodes that uses the private IP space will survive a reboot. It is only the public facing since that needs a consistent IP like the ansible tower node.

**Comment by rlopez133** [comment]
Closing this as its from 2019. If needs to be re-opened feel free to comment.
